abdiff/core/run_ab_transforms.py

```diff
@@ -1 +1,187 @@
+    )
+    return get_transformed_files(run_directory)
+
+
+def run_docker_container(
+    docker_image: str,
+    transformed_directory: str,
+    input_file: str,
+    docker_client: docker.client.DockerClient,
+) -> Container:
+    """Run transmogrifier via Docker container to transform input file."""
+    source, output_filename = parse_transform_details_from_extract_filename(input_file)
+    container = docker_client.containers.run(
+        docker_image,
+        command=[
+            f"--input-file={input_file}",
+            f"--output-file=/tmp/{output_filename}",
+            f"--source={source}",
+        ],
+        detach=True,
+        environment={
+            "AWS_ACCESS_KEY_ID": CONFIG.AWS_ACCESS_KEY_ID,
+            "AWS_SECRET_ACCESS_KEY": CONFIG.AWS_SECRET_ACCESS_KEY,
+            "AWS_SESSION_TOKEN": CONFIG.AWS_SESSION_TOKEN,
+        },
+        labels={
+            "docker_image": docker_image,
+            "source": source,
+            "input_file": input_file,
+        },
+        volumes=[f"{transformed_directory}:/tmp"],
+    )
+    logger.info(
+        f"Container '{container.id}' (Docker image: {docker_image}) "
+        f"RUNNING transform for '{source}' input_file: {input_file}."
+    )
+    return container
+
+
+def wait_for_containers(containers: list[Container], timeout: int = 180) -> None:
+    """Given a list of running containers, wait until all containers exit."""
+    exited_containers: list[Container] = []
+    start_time = time.time()
+    while len(exited_containers) < len(containers):
+        running_containers = [
+            container for container in containers if container not in exited_containers
+        ]
+        if time.time() - start_time >= timeout:
+            raise DockerContainerRuntimeExceededTimeoutError(
+                containers=containers, timeout=timeout
+            )
+        for container in running_containers:
+            container.reload()
+            if container.status == "exited":
+                logger.info(f"Container {container.id} exited.")
+                exited_containers.append(container)
+        time.sleep(0.25)
+
+
+def get_transformed_files(run_directory: str) -> tuple[list[str], ...]:
+    """Get list of filepaths to transformed JSON files.
+
+    Args:
+        run_directory (str): Run directory.
+
+    Returns:
+        tuple[list[str]]: Tuple containing lists of paths to transformed
+            JSON files for each image, relative to 'run_directory'.
+    """
+    filepaths = []
+    for root, dirs, files in os.walk(Path(run_directory) / "transformed"):
+        dirs.sort()
+        sub_filepaths = []
+        if root.endswith("transformed"):
+            # skip 'transformed' root directory
+            continue
+        for filename in files:
+            relative_dir = root.replace(run_directory, "")
+            sub_filepaths.append(os.path.join(relative_dir, filename))
+        filepaths.append(sub_filepaths)
+    return tuple(filepaths)
```

Comment:
Used `glob` in `dspace-api-python-scripts` up until our S3 conversion @jonavellecuerdo, so you could find examples in the old commits

Reply:
Given the [comment above](https://github.com/MITLibraries/transmogrifier-ab-diff/pull/6#discussion_r1755301616) about using `os.path.abspath` I realized there is a `os.path.relpath` as well!

Was curious how it would fit here:
```python
ordered_files = []
for version in ["a", "b"]:
    absolute_filepaths = glob.glob(f"{run_directory}/transformed/{version}/*.json")
    relative_filepaths = [
        os.path.relpath(file, run_directory) for file in absolute_filepaths
    ]
    ordered_files.append(relative_filepaths)
return tuple(ordered_files)
```

Interestingly, this revealed what could be a bug?  The test `XXX` failed:

```
  Full diff:
    (
        [
  -         '/transformed/a/source-2024-01-01-full-transformed-records-to-index.json',
  ?          -
  +         'transformed/a/source-2024-01-01-full-transformed-records-to-index.json',
        ],
        [
  -         '/transformed/b/source-2024-01-01-full-transformed-records-to-index.json',
  ?          -
  +         'transformed/b/source-2024-01-01-full-transformed-records-to-index.json',
        ],
    )
```

What I think this indicates is that the version as-is (that uses `os.walk()`) was including the leading slash `/`, thus making it an *absolute* path if used by anything downstream.  

While we certainly could just strip that, it feels like the use of `os.path.relpath` could be helpful in that it enforces a true relative path for a file, and not quite as much path string manipulation.

Reply:
Thank you, both! I have a better understanding of how `glob.glob` works and will remember to use `os.path.relpath` and `os.path.abspath` in the future. I updated `get_transformed_files` to follow the second suggestion from @ghukill , which uses `os.path.relpath` + `glob`.

Reply:
See [updated `get_transformed_files`](https://github.com/MITLibraries/transmogrifier-ab-diff/pull/6/files#diff-7a49d9dbf619fca67eb5ed5ed16b6c62440a4cc6898754cd835d2b819be74ddcR204-R211).